Get test data file directly in rebalance script

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
The value of `['contents']['targets']` is not always a dict. Here is an example of the data that is causing a stack trace because it is a list and has no `.items()` attribute.

```json
{
    "contents": "{\"targets\":
        [\"connection_windows_ssh\",
        \"connection_psrp\",
        \"win_become\",
        \"win_script\",
        \"win_fetch\",
        \"win_async_wrapper\",
        \"module_utils_Ansible.ModuleUtils.CommandUtil\",
        \"module_utils_Ansible.ModuleUtils.Backup\",
        \"module_utils_Ansible.Service\",
        \"windows-minimal\",
        \"module_utils_Ansible.Basic\",
        \"module_utils_Ansible.Process\",
        \"collections\",
        \"module_utils_Ansible.ModuleUtils.WebRequest\",
        \"connection_winrm\",
        \"module_utils_Ansible.ModuleUtils.CamelConversion\",
        \"win_module_utils\",
        \"binary_modules_winrm\",
        \"module_utils_Ansible.ModuleUtils.ArgvParser\",
        \"module_utils_Ansible.ModuleUtils.FileUtil\",
        \"module_utils_Ansible.ModuleUtils.SID\",
        \"win_exec_wrapper\",
        \"windows-paths\",
        \"module_utils_Ansible.Become\",
        \"win_raw\",
        \"module_utils_Ansible.ModuleUtils.AddType\",
        \"module_utils_Ansible.Privilege\",
        \"module_utils_Ansible.ModuleUtils.Legacy\",
        \"module_utils_Ansible.AccessToken\",
        \"module_utils_Ansible.ModuleUtils.LinkUtil\",
        \"module_utils_Ansible.ModuleUtils.PrivilegeUtil\",
        \"collections_relative_imports\"]
...
}
```

**Steps to reproduce:**

```
> python -VV
Python 3.8.2 (default, May 12 2020, 12:11:39)
[Clang 11.0.3 (clang-1103.0.32.59)]

> python3 download.py https://app.shippable.com/github/ansible/ansible/runs/168526 --test-results --job-number 13

> python3 rebalance.py -p ansible/ansible/168526 5
```

**Expected Results:**

Prints out rebalanced groups.

**Actual Results:**
```
Traceback (most recent call last):
  File "rebalance.py", line 181, in <module>
    main()
  File "rebalance.py", line 35, in main
    rebalance(args)
  File "rebalance.py", line 117, in rebalance
    target_times = get_raw_test_targets(args, test_path)
  File "rebalance.py", line 90, in get_raw_test_targets
    for target_name, target_info in info.get('targets', {}).items():
AttributeError: 'list' object has no attribute 'items'
```

Also add some more instructions in the comments.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`hacking/shippable/rebalance.py`

samdoran committed:
Account for list of targets

samdoran committed:
Update instructions

samdoran commented on hacking/shippable/rebalance.py:
```diff
@@ -86,15 +91,18 @@ def get_raw_test_targets(args, test_path):
-            for target_name, target_info in info.get('targets', {}).items():
-                target_runtime = int(target_info.get('run_time_seconds', 0))
+            targets = info.get('targets')
+            if isinstance(targets, dict):
```
I could also catch and handle the exception if that's better than checking the type.

jborean93 replied:
What is the type in the scenario you've seen? It sounds like `targets` is something and we shouldn't be ignoring it?

samdoran replied:
> What is the type in the scenario you've seen?

It's a list. I put an example of the data it is stack tracing on in the issue description.

>It sounds like targets is something and we shouldn't be ignoring it?

I was thinking that too. But when `targets` is a list, we don't have any of the other data we need, namely `run_time_seconds`. I didn't know what else to do besides ignore it. I suppose we could at least list the target name and just `0` or the run time, but that's an inaccurate representation.

The bigger question I have is why are we sometimes getting only a list of targets with no job data?

jborean93 commented on hacking/shippable/rebalance.py:
```diff
@@ -86,15 +91,18 @@ def get_raw_test_targets(args, test_path):
                 continue
 
             info = json.loads(test['contents'])
+            targets = info.get('targets')
```
You now have this defined twice.

samdoran replied:
Doh. 😊

samdoran committed:
Remove double var

samdoran commented:
Seems the real issue is it's looking at files it should not be. We need to be looking at the JSON file inside `data` folder and not other JSON files.

samdoran commented:
Alright, I see what's happening.

This script would look at `test.json` to find the path to `test/testresults/data/[test data].json`. It was encountering a list in a `contents['targets']` field that was causing the problem.

I'm going to just got the test results data directly and avoid processing the `test.json` file.

samdoran committed:
Get the test data directly

Rather than looking through tests.json to find the data file, look for it explicitly
within a given target to avoid problems processing data in tests.json.
